File: bonsoir_android/android/src/main/kotlin/fr/skyost/bonsoir/discovery/BonsoirServiceDiscovery.kt
package fr.skyost.bonsoir.discovery

import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.os.Handler
import android.os.Looper
import android.util.Log
import fr.skyost.bonsoir.BonsoirPlugin
import fr.skyost.bonsoir.BonsoirService
import fr.skyost.bonsoir.SuccessObject
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import java.util.concurrent.ConcurrentLinkedQueue
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Allows to find NSD services on local network.
 *
 * @param id The listener identifier.
 * @param printLogs Whether to print debug logs.
 * @param onDispose Triggered when this instance is being disposed.
 * @param nsdManager The NSD manager instance.
 * @param messenger The Flutter binary messenger.
 */
class BonsoirServiceDiscovery(
    private val id: Int,
    private val printLogs: Boolean,
    private val onDispose: Runnable,
    private val nsdManager: NsdManager,
    messenger: BinaryMessenger
) : NsdManager.DiscoveryListener, NsdManager.ResolveListener {

    /**
     * The current event channel.
     */
    private val eventChannel: EventChannel =
        EventChannel(messenger, "${BonsoirPlugin.channel}.discovery.$id")

    /**
     * The current event sink.
     */
    private var eventSink: EventChannel.EventSink? = null

    /**
     * Whether the discovery is currently active.
     */
    private var isDiscoveryActive: Boolean = false

    companion object {
        /**
         * Whether the resolver is currently busy.
         */
        private val isResolverBusy: AtomicBoolean = AtomicBoolean(false)

        /**
         * All services pending for resolution.
         */
        private val pendingServices =
            ConcurrentLinkedQueue<Pair<NsdServiceInfo, BonsoirServiceDiscovery>>()
    }

    /**
     * Initializes this instance.
     */
    init {
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink) {
                this@BonsoirServiceDiscovery.eventSink = eventSink
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
    }

    fun discoverServices(type: String) {
        nsdManager.discoverServices(type, NsdManager.PROTOCOL_DNS_SD, this)
    }

    override fun onDiscoveryStarted(regType: String) {
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir discovery started : $regType")
        }

        isDiscoveryActive = true
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(SuccessObject("discoveryStarted").toJson())
        }
    }

    override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir failed to start discovery : $errorCode")
        }

        Handler(Looper.getMainLooper()).post {
            eventSink?.error("discoveryError", "Bonsoir failed to start discovery", errorCode)
        }
        dispose(false)
    }

    override fun onServiceFound(service: NsdServiceInfo) {
        val bonsoirService = BonsoirService(service)
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir has found a service : $bonsoirService")
        }

        Handler(Looper.getMainLooper()).post {
            eventSink?.success(SuccessObject("discoveryServiceFound", bonsoirService).toJson())
        }
    }

    override fun onServiceLost(service: NsdServiceInfo) {
        val bonsoirService = BonsoirService(service)
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] A Bonsoir service has been lost : $bonsoirService")
        }

        Handler(Looper.getMainLooper()).post {
            eventSink?.success(SuccessObject("discoveryServiceLost", bonsoirService).toJson())
        }
    }

    override fun onDiscoveryStopped(serviceType: String) {
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir discovery stopped : $serviceType")
        }

        isDiscoveryActive = false
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(SuccessObject("discoveryStopped").toJson())
        }
    }

    override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
        if (printLogs) {
            Log.d(
                BonsoirPlugin.tag,
                "[$id] Bonsoir has encountered an error while stopping the discovery : $errorCode"
            )
        }

        Handler(Looper.getMainLooper()).post {
            eventSink?.error(
                "discoveryError",
                "Bonsoir has encountered an error while stopping the discovery",
                errorCode
            )
        }
        dispose()
    }

    /**
     * Resolves the given service.
     */
    fun resolveService(name: String, type: String) {
        val service = NsdServiceInfo().apply {
            serviceName = name
            serviceType = type
        }
        if (isResolverBusy.compareAndSet(false, true)) {
            forceServiceResolution(service, this)
        } else {
            pendingServices.add(Pair(service, this))
        }
    }

    /**
     * Resolves the next NSD service pending resolution.
     */
    private fun resolveNextInQueue() {
        val nextService: Pair<NsdServiceInfo, BonsoirServiceDiscovery>? = pendingServices.poll()
        if (nextService == null) {
            isResolverBusy.set(false)
        } else {
            forceServiceResolution(nextService.first, nextService.second)
        }
    }

    /**
     * Forces a service resolution.
     *
     * @param service The service.
     * @param discovery The discovery instance.
     */
    private fun forceServiceResolution(
        service: NsdServiceInfo,
        discovery: BonsoirServiceDiscovery
    ) {
        if (isDiscoveryActive) {
            nsdManager.resolveService(service, discovery)
        }
    }

    override fun onServiceResolved(service: NsdServiceInfo) {
        val bonsoirService = BonsoirService(service)
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir has resolved a service : $bonsoirService")
        }

        Handler(Looper.getMainLooper()).post {
            eventSink?.success(
                SuccessObject("discoveryServiceResolved", bonsoirService).toJson()
            )
        }
        resolveNextInQueue()
    }

    override fun onResolveFailed(service: NsdServiceInfo, errorCode: Int) {
        if (printLogs) {
            Log.d(BonsoirPlugin.tag, "[$id] Bonsoir has failed to resolve a service : $errorCode")
        }

        Handler(Looper.getMainLooper()).post {
            eventSink?.success(
                SuccessObject("discoveryServiceResolveFailed", BonsoirService(service)).toJson()
            )
        }
        resolveNextInQueue()
    }

    /**
     * Disposes the current class instance.
     */
    fun dispose(stopDiscovery: Boolean = true) {
        if (stopDiscovery && isDiscoveryActive) {
            nsdManager.stopServiceDiscovery(this)
            isDiscoveryActive = false
        }
        onDispose.run()
    }
}

File: bonsoir_darwin/darwin/Classes/Discovery/BonsoirServiceDiscovery.swift
#if canImport(Flutter)
    import Flutter
#endif
#if canImport(FlutterMacOS)
    import FlutterMacOS
#endif
import Network

/// Allows to find net services on local network.
@available(iOS 13.0, macOS 10.15, *)
class BonsoirServiceDiscovery: NSObject, FlutterStreamHandler {
    /// The delegate identifier.
    let id: Int

    /// Whether to print debug logs.
    let printLogs: Bool

    /// Triggered when this instance is being disposed.
    let onDispose: (Bool) -> Void

    /// The type we're listening to.
    let type: String
    
    /// The current browser instance.
    let browser: NWBrowser

    /// The current event channel.
    var eventChannel: FlutterEventChannel?

    /// The current event sink.
    var eventSink: FlutterEventSink?
    
    /// Contains all found services.
    var services: [BonsoirService] = []
    
    /// Contains all services wh're currently resolving.
    var resolvingServices: [DNSServiceRef?: BonsoirService] = [:]

    /// Initializes this class.
    public init(id: Int, printLogs: Bool, onDispose: @escaping (Bool) -> Void, messenger: FlutterBinaryMessenger, type: String) {
        self.id = id
        self.printLogs = printLogs
        self.onDispose = onDispose
        self.type = type
        browser = NWBrowser(for: .bonjourWithTXTRecord(type: type, domain: "local."), using: .tcp)
        super.init()
        browser.stateUpdateHandler = stateHandler
        browser.browseResultsChangedHandler = browseHandler
        eventChannel = FlutterEventChannel(name: "\(SwiftBonsoirPlugin.package).discovery.\(id)", binaryMessenger: messenger)
        eventChannel?.setStreamHandler(self)
    }
    
    /// Handles state changes.
    func stateHandler(_ newState: NWBrowser.State) {
        switch newState {
        case .ready:
            if printLogs {
                SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "Bonsoir discovery started : \(type)")
            }
            eventSink?(SuccessObject(id: "discoveryStarted").toJson())
        case .failed(let error):
            if printLogs {
                SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "Bonsoir has encountered an error during discovery : \(error)")
            }
            eventSink?(FlutterError.init(code: "discoveryError", message: "Bonsoir has encountered an error during discovery.", details: error))
            dispose()
        case .cancelled:
            if printLogs {
                SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "Bonsoir discovery stopped : \(type)")
            }
            eventSink?(SuccessObject(id: "discoveryStopped").toJson())
            dispose()
        default:
            break
        }
    }

    /// Handles the browsing of services.
    func browseHandler(_ newResults: Set<NWBrowser.Result>, _ changes: Set<NWBrowser.Result.Change>) {
        for change in changes {
            switch change {
            case .added(let result):
                if case .service(let name, let type, _, _) = result.endpoint {
                    let service = BonsoirService(name: name, type: type, port: 0, host: nil, attributes: nil)
                    if case .bonjour(let records) = result.metadata {
                        service.attributes = records.dictionary
                    }
                    if printLogs {
                        SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "Bonsoir has found a service : \(service.description)")
                    }
                    eventSink?(SuccessObject(id: "discoveryServiceFound", service: service).toJson())
                    services.append(service)
                }
            case .removed(let result):
                if case .service(let name, let type, _, _) = result.endpoint {
                    let service = services.first(where: {$0.name == name && $0.type == type})
                    if service == nil {
                        break
                    }
                    if printLogs {
                        SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "A Bonsoir service has been lost : \(service!.description)")
                    }
                    eventSink?(SuccessObject(id: "discoveryServiceLost", service: service).toJson())
                    if let index = self.services.firstIndex(where: { $0 === service }) {
                        self.services.remove(at: index)
                    }
                }
            case .changed(let old, let new, _):
                if case .service(let newName, let newType, _, _) = new.endpoint {
                    if case .service(let oldName, let oldType, _, _) = old.endpoint {
                        let service = services.first(where: {$0.name == oldName && $0.type == oldType})
                        if service == nil {
                            break
                        }
                        if printLogs {
                            SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "A Bonsoir service has changed : \(service!.description)")
                        }
                        eventSink?(SuccessObject(id: "discoveryServiceLost", service: service).toJson())
                        service!.name = newName
                        service!.type = newType
                        if case .bonjour(let newRecords) = new.metadata {
                            service!.attributes = newRecords.dictionary
                        }
                        eventSink?(SuccessObject(id: "discoveryServiceFound", service: service).toJson())
                    }
                }
            default:
                break
            }
        }
    }

    func onListen(withArguments arguments: Any?, eventSink: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = eventSink
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
    
    /// Resolves a service.
    public func resolveService(name: String, type: String) -> Bool {
        for result in browser.browseResults {
            if case .service(let name, let type, _, _) = result.endpoint {
                let service = services.first(where: {$0.name == name && $0.type == type})
                if service == nil {
                    return false
                }
                var sdRef: DNSServiceRef? = nil
                let error = DNSServiceResolve(&sdRef, 0, 0, name, type, "local.", { (sdRef, flags, interfaceIndex, errorCode, fullName, hosttarget, port, txtLen, txtRecord, context) in
                    let discovery = Unmanaged<BonsoirServiceDiscovery>.fromOpaque(context!).takeUnretainedValue()
                    let service = discovery.resolvingServices[sdRef]
                    if errorCode == kDNSServiceErr_NoError {
                        if hosttarget != nil {
                            service!.host = String(cString: hosttarget!)
                        }
                        service!.port = Int(port)
                        if discovery.printLogs == true {
                            SwiftBonsoirPlugin.log(category: "discovery", id: discovery.id, message: "Bonsoir has resolved a service : \(service!.description)")
                        }
                        discovery.eventSink?(SuccessObject(id: "discoveryServiceResolved", service: service).toJson())
                    } else {
                        if discovery.printLogs {
                            SwiftBonsoirPlugin.log(category: "discovery", id: discovery.id, message: "Bonsoir has failed to resolve a service : \(errorCode)")
                        }
                        discovery.stopResolution(sdRef: sdRef, remove: sdRef != nil)
                        discovery.eventSink?(SuccessObject(id: "discoveryServiceResolveFailed", service: service).toJson())
                    }
                }, Unmanaged.passUnretained(self).toOpaque())
                if error == kDNSServiceErr_NoError {
                    resolvingServices[sdRef] = service
                    DNSServiceProcessResult(sdRef)
                } else {
                    if printLogs {
                        SwiftBonsoirPlugin.log(category: "discovery", id: id, message: "Bonsoir has failed to resolve a service : \(error)")
                    }
                    stopResolution(sdRef: sdRef, remove: false)
                    eventSink?(SuccessObject(id: "discoveryServiceResolveFailed", service: service).toJson())
                }
                return true
            }
        }
        return false
    }
    
    /// Stops the resolution of the given service.
    private func stopResolution(sdRef: DNSServiceRef?, remove: Bool = true) {
        if remove {
            resolvingServices.removeValue(forKey: sdRef)
        }
        DNSServiceRefDeallocate(sdRef)
    }
    
    /// Starts the discovery.
    public func start() {
        browser.start(queue: .main)
    }

    /// Disposes the current class instance.
    public func dispose() {
        for sdRef in resolvingServices.keys {
            stopResolution(sdRef: sdRef, remove: false)
        }
        resolvingServices.removeAll()
        services.removeAll()
        if case .setup || .ready = browser.state {
            browser.cancel()
        }
        onDispose()
    }
}
